### PR TITLE
ExpandDim-DynamicBroadcastInDim simplifier enhancement.

### PR DESCRIPTION
**Pull request by JamesTheZ**
Enhance the conversion from "ExpandDim - DynamicBroadcastInDim" to DynamicBroadcastInDim.

### CODE REVIEW

File: tao_compiler/mlir/disc/transforms/disc_shape_simplifier.cc [hunk 1]
```diff
@@ -202,7 +202,12 @@ class DynamicReshapeOpShapeInference
   }
 };
 
-// Match following patterns:
+// Convert an expand-like reshape op followed by a dynamic-broadcast-in-dim op
+// to a single dynamic-broadcast-in-dim op. The dim size of expanded dims should
+// be one.
+//
+// An example is like following:
```
Comment (linearhit) [inline tao_compiler/mlir/disc/transforms/disc_shape_simplifier.cc:209]: From the code, we are looking for a DynamicReshapeOp or ReshapeOp to match the pattern. But i didn't see any reshape here in the example?

File: tao_compiler/mlir/disc/transforms/disc_shape_simplifier.cc [hunk 2]
```diff
@@ -212,56 +217,130 @@ class DynamicReshapeOpShapeInference
...
+    } else {
+      auto fromElementsOp = dyn_cast_or_null<tensor::FromElementsOp>(
+          dynReshapeOp->getOperand(1).getDefiningOp());
+      if (!fromElementsOp) {
+        return failure();
+      }
+
+      for (auto dim : fromElementsOp->getOperands()) {
+        auto staticDim =
+            dyn_cast_or_null<arith::ConstantIndexOp>(dim.getDefiningOp());
```
Comment (wyzero) [inline tao_compiler/mlir/disc/transforms/disc_shape_simplifier.cc:280]: What about `arith::ConstantIntOp`?

File: tao_compiler/mlir/disc/transforms/disc_shape_simplifier.cc (hunk 1 above)
Comment (JamesTheZ) [reply tao_compiler/mlir/disc/transforms/disc_shape_simplifier.cc:209]: This example only shows a dynamic-reshape based pattern. ReshapeOp based pattern is similar except that the batch-dim value is constant folded.

**Review by linearhit — APPROVED** [review]
LGTM

File: tao_compiler/mlir/disc/transforms/disc_shape_simplifier.cc (hunk 2 above)
Comment (JamesTheZ) [reply tao_compiler/mlir/disc/transforms/disc_shape_simplifier.cc]: Thanks for the comment. Fixed.

**Pull request merged**